docs/design.md

```diff
@@ -0,0 +1,257 @@
+        Previous        AdvertisementID
+        Provider        peer.ID
+        Signature       []byte
+        GraphSupport    bool
+        
+}
+```
+- `ID` is the unique identifier used by the data provider to identify this update.
+- `Previous` links the current advertisement for the provider with the previous one. This allows
+providers to track the announcements they are broadcasting, and 
+indexer nodes to realize when they have missed a previous advertisement.
+- `Provider` determines the source of the advertisement.
+- `Signature` is a signature over `ID` and `Previous` to authenticate that the advertisement has been generated by `Provider`.
+- `GraphSupport` is used by providers to signal indexer nodes that they support IPLD-aware ingestion sessions (see [protocol](#protocol)).
+
+`AdvertisementID` can be chosen by data providers in the way that suit them best according to their internal
+data indexing strategy. The only requirement for
+`AdvertisementID` is that it should allow providers to uniquely identify the batch of data, or data update
+they are referring to in the advertisement. Indexer nodes use these IDs in their ingestion requests to fetch indexing
+data from the provider.
+
+For instance, a Filecoin miner may choose to use as `AdvertisementID` a combination of a `dealID` with a prefix to identify if
+the advertisement points to a new deal or the expiration of a deal. Thus, when a miner receives an ingestion
+request for `AdvertisementID` from an indexer node, it can easily build the ingestion response (either with the
+batch of new CIDs provided in the deal, or with the list of CIDs the indexer should stop providing due to a deal expiration). Other data providers may choose to generate their `AdvertisementID`s in alternative ways.
+
+For indexer nodes, `AdvertisementID` is an opaque `String` and has no additional significance apart from the unique ID
+for a data provider advertisement.
+
+```go
+type AdvertisementID String
+```
+<!-- We may not need to include `Previous` in Advertisements, as they can be signalled in Ingestion responses. When an advertisement is seen, a request is sent to providers, and providers can signal the previous one. Then the indexer chooses if it is already in sync or not. However, chaining advertisements comes pretty handy for data providers, as it gives them a way to identify the sequence of updates and serve ingestion requests -->
+
+### Ingestion Request
+When an indexer node finds an unseen `Advertisement` and realizes that it is not in sync with a data provider, it sends
+an ingestion request to the data provider to index/remove data.
+```go
+type IngestionRequest struct {
+        ID              AdvertisementID
+        Start           uint64
+        Count           uint64
+}
+```
+- `ID`: `AdvertisementID` for which data is being requested.
+- `Start`: Starting entry from which data wants to be requested.
+- `Count`: Maximum number of entries that want to be received in the response.
```

Comment:
is this for pagination within the advertisement?
The indexer won't know the total count though, so how does it know when it's gotten all?
Isn't the common case 'i want all of the advertisement'? can we defer pagination for now?

Reply:
Yes. This is for pagination within the advertisement. `Count` is only used by the indexer node to limit the maximum amount of entries it wants to receive in a single Response.
The provider signals the total number of entries of the advertisement in `Size`, so after the first Response, the indexer gets knowledge of the total number of entries of the advertisement. 

I personally think we should aim to include pagination in the MVP, because the amount of data to be exchanged between provider an indexer may be quite large even for the MVP, and we can easily reach the `network.MessageSizeMax` soft limit if we choose to implement a message-based protocol (but happy to be convinced otherwise :) )

Reply:
i'd like to have the concrete transport / protocol considered jointly with this if possible. I feel like there's a range of options and am not convinced pagination like this will be easiest to implement.

If i'm doing a bulk HTTP-style upload, I probably don't care too much about size, the http layer will let me push the whole list of CIDs in one go, and potentially do range queries over the record if partial gaps need to be filled. Having it be non-dynamic means the provider can generate the 'announcement' once and leave it as a static json file on s3 or similar, rather than needing a dynamic endpoint to handle potentially-dynamic paging.

If i'm doing a libp2p-style protocol, we have bitswap and graphsync already for transferring dags. can't we use one of those rather than re-inventing a chunking method here?

Maybe the underlying thing to figure out is: I think it's never going to the case that I want to partially fetch an announcement. in order to verify it, i'm either going to want all of it, or none of it as the granularity that i do rate limiting on as an indexer node.

Reply:
> Having it be non-dynamic means the provider can generate the 'announcement' once and leave it as a static json file on s3 or similar, rather than needing a dynamic endpoint to handle potentially-dynamic paging.

This should definitely be a requirement. Announcement should be static. And ideally the ingestion protocol should be stateless to allow any kind of use case.

> If i'm doing a libp2p-style protocol, we have bitswap and graphsync already for transferring dags. can't we use one of those rather than re-inventing a chunking method here?

Agree! As I mention by the end, ideally if the provider represent data internally using IPLD, we may default to Graphsync, but I wasn't sure if we could expect this for all providers, specially because for the initial design we were considering the use of pagination already. 

> I think it's never going to the case that I want to partially fetch an announcement. in order to verify it, i'm either going to want all of it, or none of it as the granularity that i do rate limiting on as an indexer node.

Right. Ingestions are going to be atomic. You'll want to fetch all or nothing.

Thanks for the feedback. I'll do another pass addressing this.